File: src/modules/algorithm/seed/algorithms/knapsack-fractional/javascript/greedy-solution.js
/**
 * Implements the fractional knapsack algorithm using a greedy approach.
 *
 * @param {number[]} weights - An array of item weights
 * @param {number[]} values - An array of item values
 * @param {number} capacity - The knapsack capacity
 * @returns {number} The maximum value that can be carried in the knapsack
 *
 * Time Complexity: O(n log n) due to sorting
 */
function fractionalKnapsack(weights, values, capacity) {
  const n = weights.length;
  if (n === 0) return 0;

  // Create items array with value-to-weight ratio
  const items = weights.map((weight, index) => ({
    index,
    ratio: values[index] / weight,
  }));

  // Sort items by value-to-weight ratio in descending order
  items.sort((a, b) => b.ratio - a.ratio);

  let totalValue = 0;
  let remainingCapacity = capacity;

  for (const item of items) {
    const weight = weights[item.index];
    const value = values[item.index];

    if (remainingCapacity >= weight) {
      // Take the whole item
      totalValue += value;
      remainingCapacity -= weight;
    } else {
      // Take a fraction of the item
      const fraction = remainingCapacity / weight;
      totalValue += value * fraction;
      break; // Knapsack is full
    }
  }

  return totalValue;
}

module.exports = { fractionalKnapsack };

File: src/modules/algorithm/seed/algorithms/knapsack-fractional/python/greedy-solution.py
from typing import List

def fractional_knapsack(weights: List[float], values: List[float], capacity: float) -> float:
    """
    Implements the fractional knapsack algorithm using a greedy approach.

    Args:
        weights (List[float]): A list of item weights.
        values (List[float]): A list of item values.
        capacity (float): The knapsack capacity.

    Returns:
        float: The maximum value that can be carried in the knapsack.

    Time Complexity: O(n log n) due to sorting
    """
    n = len(weights)
    if n == 0:
        return 0.0

    # Calculate value-to-weight ratio for each item
    items = [(values[i] / weights[i], weights[i], values[i]) for i in range(n)]

    # Sort items by value-to-weight ratio in descending order
    items.sort(reverse=True)

    total_value = 0.0
    remaining_capacity = capacity

    for ratio, weight, value in items:
        if remaining_capacity >= weight:
            # Take the whole item
            total_value += value
            remaining_capacity -= weight
        else:
            # Take a fraction of the item
            fraction = remaining_capacity / weight
            total_value += value * fraction
            break  # Knapsack is full

    return total_value 